### PR TITLE
feat: support date parts

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #3185 
Closes: #3326

### Description:

Screenshot 2022-09-27 at 19 06 38

All options:
              - RAW
              - MILLISECOND
              - SECOND
              - MINUTE
              - HOUR
              - DAY
              - DAY_OF_WEEK_INDEX **( new )** 
              - DAY_OF_WEEK_NAME **( new )** 
              - DAY_OF_MONTH_NUM **( new )** 
              - DAY_OF_YEAR_NUM **( new )** 
              - WEEK
              - MONTH
              - MONTH_NUM **( new )** 
              - MONTH_NAME **( new )** 
              - QUARTER
              - QUARTER_NUM **( new )** 
              - QUARTER_NAME **( new )** 
              - YEAR
              - YEAR_NUM **( new )** 

Options that don't have a suffix ( aka all the existing values ) return date values. 
Options that end with `_INDEX` or `_NUM` return numeric values. 
Options that end with `_NAME` return string values. 

Year style/format will be fixed in a separate PR as that is a new feature request.  #3325

### CODE REVIEW

**Comment by TuringLovesDeathMetal** [comment]
wauuuuuw!!! Looks amazing @ZeRego 😍 really cool to see all of the new options!

I can't seem to interact with them in the demo project 🤔 but maybe that's because it doesn't work in the demo deployment with new fields? 

Anyways, the only bit of feedback I'd have is: can we set the field names to match the names of the intervals? 

E.g. 

`MONTH_NUM` would appear like `month num` in the sidebar. 

I find it a bit confusing to have them all named `month` in the sidebar (even though they all have different value icons 👀 )

Especially when I'm searching for filters/values - I have to read the name, then read the icon which feels like a recipe for people "not noticing the icon"

**Comment by ZeRego** [comment]
> I can't seem to interact with them in the demo project

@TuringLovesDeathMetal I need to update the demo repo so the new options are available in the PR env and e2e tests can pass. I'm only going to update that repo once we agree on the new time interval options.

> The only bit of feedback I'd have is: can we set the field names to match the names of the intervals?
> I find it a bit confusing to have them all named month in the sidebar (even though they all have different value icons 👀 )

Yes, we can make that change. Waiting for more feedback before I change it. 
My initial thought was that have `Month number` and `Month name` would be to verbose and users wouldn't have all the options available at the same time.

**Comment by TuringLovesDeathMetal** [comment]
@ZeRego do `YEAR` and `YEAR_NUM` return different values?

**Comment by ZeRego** [comment]
> @ZeRego do `YEAR` and `YEAR_NUM` return different values?

All the "basic" options (day, week, month, quarter, year) return a Date value

**Comment by TuringLovesDeathMetal** [comment]
hmmm...okay, can we make them print out like a number (i.e. not change the format that we currently have in the app?)

Otherwise, can we make the `default` time intervals look the same as they currently do somehow by using some of the other time interval options? 🤔 

image

**Comment by ZeRego** [comment]
> hmmm...okay, can we make them print out like a number (i.e. not change the format that we currently have in the app?)
> 
> Otherwise, can we make the `default` time intervals look the same as they currently do somehow by using some of the other time interval options? 🤔
> 
> image

Date values will keep the existing formatting in the results table.

**Comment by TuringLovesDeathMetal** [comment]
Okay, cool. 

Then I'd suggest we remove `YEAR_NUM` from the format options!

**Comment by ZeRego** [comment]
> Okay, cool.
> 
> Then I'd suggest we remove `YEAR_NUM` from the interval options!

why ? isn't it as useful as month_num or day_of_month_num ?  
We don't loose anything in supporting it.

**Comment by TuringLovesDeathMetal** [comment]
ya, I guess 🤷 it just returns the same thing as `YEAR_NUM` in the UI, so I would be a bit confused. 

We can just write good docs for it :)

File: packages/common/src/utils/timeFrames.ts [hunk 1]
```diff
@@ -1,41 +1,338 @@
...
+        timeFrame: TimeFrames,
+        originalSql: string,
+        type: DimensionType,
+    ) => {
+        // https://cloud.google.com/bigquery/docs/reference/standard-sql/format-elements#format_elements_date_time
+        const timeFrameExpressions: Record<TimeFrames, string | null> = {
+            ...nullTimeFrameMap,
+            [TimeFrames.DAY_OF_WEEK_NAME]: '%A',
+            [TimeFrames.MONTH_NAME]: '%B',
+            [TimeFrames.QUARTER_NAME]: 'Q%Q',
```
Comment (rephus) [inline packages/common/src/utils/timeFrames.ts:84]: `Q%Q` or `%Q` ?

File: packages/e2e/cypress/e2e/createProjects.cy.ts [hunk 2]
```diff
@@ -119,6 +119,72 @@ const testQuery = () => {
     cy.findByText('Unique order count').click();
 };
 
+const testTimeIntervalsResults = () => {
```
Comment (rephus) [inline packages/e2e/cypress/e2e/createProjects.cy.ts:122]: :raised_hands:

**Comment by rephus** [comment]
Should we start on Sunday ? Is it consistent across different warehouses / locales ? 
image

**Comment by rephus** [comment]
is it intentional that we don't show stuff between brackets on column names ? This makes all column look the same 
image

**Comment by rephus** [comment]
Would it be useful to have a quarter-month that returns a `year-month` date (eg: 2020-q3 -> 2020-05-01`)  like what we do with weeks ? Not sure if it is going to be useful for filtering. Not a blocker though.

**Comment by rephus** [comment]
I got this error on one bigquery project, any idea why ? 

https://lightdash-pr-3316.onrender.com/projects/16a4bcbd-eb02-449b-9cac-43c2f3e9beb9/tables/events

image

**Comment by rephus** [comment]
image
 on snowflake 

https://lightdash-pr-3316.onrender.com/projects/4c7364c3-95bb-4140-8fa0-d241c9224e13/tables/events?create_saved_chart_version=%7B"tableName"%3A"events"%2C"metricQuery"%3A%7B"dimensions"%3A%5B"events_timestamp_tz_raw"%2C"events_timestamp_tz_millisecond"%2C"events_timestamp_tz_second"%5D%2C"metrics"%3A%5B%5D%2C"filters"%3A%7B%7D%2C"sorts"%3A%5B%7B"fieldId"%3A"events_timestamp_tz_raw"%2C"descending"%3Atrue%7D%5D%2C"limit"%3A500%2C"tableCalculations"%3A%5B%5D%2C"additionalMetrics"%3A%5B%5D%7D%2C"tableConfig"%3A%7B"columnOrder"%3A%5B"events_timestamp_tz_raw"%2C"events_timestamp_tz_millisecond"%2C"events_timestamp_tz_second"%2C"events_timestamp_tz_minute"%2C"events_timestamp_tz_hour"%2C"events_timestamp_tz_day"%2C"events_timestamp_tz_day_of_week_index"%2C"events_timestamp_tz_day_of_week_name"%2C"events_timestamp_tz_day_of_year_num"%2C"events_timestamp_tz_day_of_month_num"%2C"events_timestamp_tz_week"%2C"events_timestamp_tz_month"%2C"events_timestamp_tz_month_num"%2C"events_timestamp_tz_month_name"%2C"events_timestamp_tz_quarter"%2C"events_timestamp_tz_quarter_num"%2C"events_timestamp_tz_quarter_name"%2C"events_timestamp_tz_year"%2C"events_timestamp_tz_year_num"%5D%7D%2C"chartConfig"%3A%7B"type"%3A"cartesian"%2C"config"%3A%7B"layout"%3A%7B"xField"%3A"events_timestamp_tz_raw"%2C"yField"%3A%5B"events_timestamp_tz_millisecond"%5D%7D%2C"eChartsConfig"%3A%7B"series"%3A%5B%7B"encode"%3A%7B"xRef"%3A%7B"field"%3A"events_timestamp_tz_raw"%7D%2C"yRef"%3A%7B"field"%3A"events_timestamp_tz_millisecond"%7D%7D%2C"type"%3A"bar"%7D%5D%7D%7D%7D%7D

File: packages/common/src/utils/timeFrames.ts (hunk 1 above)
Comment (ZeRego) [reply packages/common/src/utils/timeFrames.ts:90]: %Q returns 3
Q%Q returns Q3

**Comment by ZeRego** [comment]
> Should we start on Sunday ? Is it consistent across different warehouses / locales ?

Added note in docs about start of the week and value ranges being different for each warehouse. 
We are not going to standardise these values right now.

> is it intentional that we don't show stuff between brackets on column names ? This makes all column look the same

We have updated labels to have `(number)` and `(name)`. The new labels are only available in new or refreshed projects in the PR env. 

> Would it be useful to have a quarter-month that returns a year-month date (eg: 2020-q3 -> 2020-05-01`) like what we do with weeks ? Not sure if it is going to be useful for filtering. Not a blocker though.

Yes, we decided to leave that one out for now as it didn't fit the name convention and because the only use case we know of for that can be solved in a different way. The use case is for the chart axis to be a categorical using the format you mentioned, when using the time interval QUARTER, which we can solve in the echart configuration.

**Comment by rephus** [comment]
I got another error on snowflake

Screenshot from 2022-09-29 16-31-02


https://lightdash-pr-3316.onrender.com/projects/4c7364c3-95bb-4140-8fa0-d241c9224e13/tables/events?create_saved_chart_version=%7B"tableName"%3A"events"%2C"metricQuery"%3A%7B"dimensions"%3A%5B"events_timestamp_tz_raw"%2C"events_timestamp_tz_millisecond"%2C"events_timestamp_tz_second"%5D%2C"metrics"%3A%5B%5D%2C"filters"%3A%7B%7D%2C"sorts"%3A%5B%7B"fieldId"%3A"events_timestamp_tz_raw"%2C"descending"%3Atrue%7D%5D%2C"limit"%3A500%2C"tableCalculations"%3A%5B%5D%2C"additionalMetrics"%3A%5B%5D%7D%2C"tableConfig"%3A%7B"columnOrder"%3A%5B"events_timestamp_tz_raw"%2C"events_timestamp_tz_millisecond"%2C"events_timestamp_tz_second"%2C"events_timestamp_tz_minute"%2C"events_timestamp_tz_hour"%2C"events_timestamp_tz_day"%2C"events_timestamp_tz_day_of_week_index"%2C"events_timestamp_tz_day_of_week_name"%2C"events_timestamp_tz_day_of_month_num"%2C"events_timestamp_tz_week"%2C"events_timestamp_tz_day_of_year_num"%2C"events_timestamp_tz_month"%2C"events_timestamp_tz_month_num"%2C"events_timestamp_tz_month_name"%2C"events_timestamp_tz_quarter"%2C"events_timestamp_tz_quarter_num"%2C"events_timestamp_tz_quarter_name"%2C"events_timestamp_tz_year"%2C"events_timestamp_tz_year_num"%5D%7D%2C"chartConfig"%3A%7B"type"%3A"cartesian"%2C"config"%3A%7B"layout"%3A%7B"xField"%3A"events_timestamp_tz_raw"%2C"yField"%3A%5B"events_timestamp_tz_millisecond"%5D%7D%2C"eChartsConfig"%3A%7B"series"%3A%5B%7B"encode"%3A%7B"xRef"%3A%7B"field"%3A"events_timestamp_tz_raw"%7D%2C"yRef"%3A%7B"field"%3A"events_timestamp_tz_millisecond"%7D%7D%2C"type"%3A"bar"%7D%5D%7D%7D%7D%7D
https://lightdash-pr-3316.onrender.com/projects/4c7364c3-95bb-4140-8fa0-d241c9224e13/tables/events?create_saved_chart_version=%7B"tableName"%3A"events"%2C"metricQuery"%3A%7B"dimensions"%3A%5B"events_timestamp_tz_raw"%2C"events_timestamp_tz_millisecond"%2C"events_timestamp_tz_second"%5D%2C"metrics"%3A%5B%5D%2C"filters"%3A%7B%7D%2C"sorts"%3A%5B%7B"fieldId"%3A"events_timestamp_tz_raw"%2C"descending"%3Atrue%7D%5D%2C"limit"%3A500%2C"tableCalculations"%3A%5B%5D%2C"additionalMetrics"%3A%5B%5D%7D%2C"tableConfig"%3A%7B"columnOrder"%3A%5B"events_timestamp_tz_raw"%2C"events_timestamp_tz_millisecond"%2C"events_timestamp_tz_second"%2C"events_timestamp_tz_minute"%2C"events_timestamp_tz_hour"%2C"events_timestamp_tz_day"%2C"events_timestamp_tz_day_of_week_index"%2C"events_timestamp_tz_day_of_week_name"%2C"events_timestamp_tz_day_of_month_num"%2C"events_timestamp_tz_week"%2C"events_timestamp_tz_day_of_year_num"%2C"events_timestamp_tz_month"%2C"events_timestamp_tz_month_num"%2C"events_timestamp_tz_month_name"%2C"events_timestamp_tz_quarter"%2C"events_timestamp_tz_quarter_num"%2C"events_timestamp_tz_quarter_name"%2C"events_timestamp_tz_year"%2C"events_timestamp_tz_year_num"%5D%7D%2C"chartConfig"%3A%7B"type"%3A"cartesian"%2C"config"%3A%7B"layout"%3A%7B"xField"%3A"events_timestamp_tz_raw"%2C"yField"%3A%5B"events_timestamp_tz_millisecond"%5D%7D%2C"eChartsConfig"%3A%7B"series"%3A%5B%7B"encode"%3A%7B"xRef"%3A%7B"field"%3A"events_timestamp_tz_raw"%7D%2C"yRef"%3A%7B"field"%3A"events_timestamp_tz_millisecond"%7D%7D%2C"type"%3A"bar"%7D%5D%7D%7D%7D%7D

**Comment by rephus** [comment]
Because quarter is supposed to be a date, it doesn't allow me to manually introduce `2020-Q3` , but shows me a datetime picker instead. Can be done on a separate PR 

image

**Comment by rephus** [comment]
I don't think we support Milliseconds on filters (tested on snowflake). Not a blocker

image

image

**Comment by rephus** [comment]
Why did I get only these 2 results on databricks ? 

image

**Review by rephus — APPROVED** [review]
Looks great to me, everything works as expected. Amazing job ! :tada:

**Pull request merged**